# Open Problems in Single-Cell Analysis

![GitHub Workflow Status](https://img.shields.io/github/workflow/status/singlecellopenproblems/singlecellopenproblems/Run%20Tests/master?label=Github%20Actions)
![Coverage Status](https://img.shields.io/codecov/c/github/singlecellopenproblems/SingleCellOpenProblems/master)
[![Netlify Status](https://api.netlify.com/api/v1/badges/83b92388-53c7-4fef-9003-e14d94c6ac6f/deploy-status)](https://app.netlify.com/sites/openproblems/deploys)
[![Code Style: Black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![Style Guide: OpenStack](https://img.shields.io/badge/style%20guide-openstack-eb1a32.svg)](https://docs.openstack.org/hacking/latest/user/hacking.html#styleguide)
[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)](https://github.com/pre-commit/pre-commit)
[![Video](https://img.shields.io/static/v1?label=YouTube&message=Visit%20channel&color=red&logo=youtube)](https://www.youtube.com/channel/UCJpqxlzxRamcA3Pv3KlYZHg)

Formalizing and benchmarking open problems in single-cell genomics.

[**Visit the Open Problems Website.**](https://openproblems.bio/)


## Guides
* For contributing guidelines, see [CONTRIBUTING.md](CONTRIBUTING.md)  
* For instructions on editing Docker images, see [docker/README.md](docker/README.md)  
* For a walkthrough of the GitHub Actions workflows and AWS Sagemaker, see [SAGEMAKER.md](SAGEMAKER.md)  
* For a description of existing an proposed tasks, see [TASKS.md](TASKS.md)  
* For a video introduction to this GitHub repository, watch our [Repository introduction](https://www.youtube.com/watch?v=tHempZCdXyA)
* For a video tutorial on adding new tasks, watch our [How to add a new task tutorial](https://www.youtube.com/watch?v=tgVG3Hp6mBc)

## The team

**Core**:
* Scott Gigante (@scottgigante)
* Daniel Burkhardt (@dburkhardt)
* Malte Luecken (@LuckyMD)
* Angela Pisco (@aopisco)
* Olga Botvinnik (@olgabot)

**Task authors** (_alphabetically_):
* Emma Dann (@emdann) - Differential Abundance 
* Wes Lewis (@weslewis) - Differential Abundance and Data Denoising
* Mohammad Lotfallahi (@M0hammadL) - Label projection task
* Qian Qin (@qinqian) - Predicting gene expression from chromatin accessibility
* Daniel Strobl (@danielStrobl) - Batch integration
* Michael Vinyard (@mvinyard) - Stress preservation in Dimensionality Reduction
* Florian Wagner (@flo-compbio) - Data denoising

**Supervision** (_alphabetically_):
* Smita Krishnaswamy, Yale
* Fabian Theis, Helmholtz Munich

**Chan Zuckerberg Initiative Support** (_alphabetically_):
* Jonah Cool
* Fiona Griffin
